### ISSUE TITLE
<w:tblGrid> seems to be optional (word 2007 file)

### ISSUE THREAD

**Issue by xsduan** [python-openxml/python-docx#547]
Sorry if this has already been brought up, I haven't been able to find a similar issue.

```xml
<w:tbl>
  <w:tblPr>
    <!-- tbl info snip -->
  </w:tblPr>
  <w:tr>
    <!-- row snip -->
  </w:tr>
  <!-- more table rows-->
</w:tbl>
```

![tblgrid](https://user-images.githubusercontent.com/9505910/45503254-2f573d00-b73b-11e8-997f-cf24a44b6b40.png)
(I changed the text but that's essentially what it looks like)

However, I keep getting an error complaining that `<w:tblGrid>` does not exist in the table. Maybe Word is error correcting or something?

Word 2007 file, running on Word 2016.

**Comment by xsduan** [comment]
```py
# monkey patch python-docx
# -----

from docx.oxml.table import CT_Tbl
from docx.oxml.exceptions import InvalidXmlError
from docx.oxml.ns import qn


@property
def col_count(self):
    """
    The number of grid columns in this table.
    """
    try:
        return len(self.tblGrid.gridCol_lst)
    except InvalidXmlError:
        # if no tblGrid, make a best guess of column count
        col_counts = set(len(tr.tc_lst) for tr in self.findall(qn('w:tr')))
        if len(col_counts) > 1:
            raise  # no w:tblGrid + variable-cell table = actual error
        return list(col_counts)[0]


CT_Tbl.col_count = col_count

# -----
# end monkey patch python-docx
```

here's a quick monkey patch I've written up, it seems to work but I don't know how you want to handle this (if at all) so I'll just leave this here

**Comment by scanny** [comment]
@xsduan What is the provenance of the .docx files you're working with? The WordprocessingML specs the `w:tblGrid` element as a one-and-only-one child of `w:tbl`:  
https://github.com/python-openxml/python-docx/blob/master/ref/xsd/wml.xsd#L2531-L2537

So it would appear your table is non-compliant, even if perhaps Word allows it.

**Comment by xsduan** [comment]
I know it’s non-compliant, but since the name of the package is docx I thought it might be best to emulate what word does & accepts (since apparently word is ever so slightly different from open office anyways).

The app that I got the docx from is using phpdocx 2.4 and running php 5.3, so it may be…a little outdated. 

Anyways, for the regular user I think it would be nice to just insert an appropriate tblGrid if it can deduce that the table is “simple” and just output a warning instead of an error.

Ugh, if it actually errored out in Word I could use that as an excuse to fix some things, but…nope ;(

Anyways, thanks.

**Comment by scanny** [comment]
Yeah, I don't think we're going to be rushing to build-in a workaround for a gap in phpdocx :)

If it were LibreOffice we might consider it, but given that the spec is perfectly clear and Word is compliant (if ever so compassionately compatible in this case), I don't think this is something we could build in. The code for tables is complex as it is and I don't think it would be good engineering to stress that with something like this.

Also, I can't see anyone with the right capabilities picking up this work. There would be a lot more to it than just popping in an empty `w:tblGrid` element. You'd have to parse the table and work out how many grid columns there were (taking merged cells into account) and infer their widths. Not a trivial problem in the general case.

**Comment by DigantDixit** [comment]
Bro

> ```python
> # monkey patch python-docx
> # -----
> 
> from docx.oxml.table import CT_Tbl
> from docx.oxml.exceptions import InvalidXmlError
> from docx.oxml.ns import qn
> 
> 
> @property
> def col_count(self):
>     """
>     The number of grid columns in this table.
>     """
>     try:
>         return len(self.tblGrid.gridCol_lst)
>     except InvalidXmlError:
>         # if no tblGrid, make a best guess of column count
>         col_counts = set(len(tr.tc_lst) for tr in self.findall(qn('w:tr')))
>         if len(col_counts) > 1:
>             raise  # no w:tblGrid + variable-cell table = actual error
>         return list(col_counts)[0]
> 
> 
> CT_Tbl.col_count = col_count
> 
> # -----
> # end monkey patch python-docx
> ```
> 
> here's a quick monkey patch I've written up, it seems to work but I don't know how you want to handle this (if at all) so I'll just leave this here

Thanks you so much bro this code saved my life
